runtime/third-party/emyg/emyg_dtoa.c

```diff
@@ -0,0 +1,452 @@
+		  641,   667,   694,   720,   747,   774,   800,   827,   853,   880,
+		  907,   933,   960,   986,  1013,  1039,  1066
+	};
+
+	//int k = (int )(ceil((-61 - e) * 0.30102999566398114)) + 374;
+	double dk = (-61 - e) * 0.30102999566398114 + 347;	// dk must be positive, so can do ceiling in positive
+	int k = (int )(dk);
+	if (k != dk)
+		k++;
+
+	unsigned index = (unsigned )((k >> 3) + 1);
+	*K = -(-348 + (int )(index << 3));	// decimal exponent no need lookup table
+
+	assert(index < sizeof(kCachedPowers_F) / sizeof(kCachedPowers_F[0]));
+	return DiyFp_from_parts(kCachedPowers_F[index], kCachedPowers_E[index]);
+}
+
+static inline void GrisuRound(char* buffer, int len, uint64_t delta, uint64_t rest, uint64_t ten_kappa, uint64_t wp_w) {
+	while (rest < wp_w && delta - rest >= ten_kappa &&
+		   (rest + ten_kappa < wp_w ||  /// closer
+			wp_w - rest > rest + ten_kappa - wp_w)) {
+		buffer[len - 1]--;
+		rest += ten_kappa;
+	}
+}
+
+static inline unsigned CountDecimalDigit32(uint32_t n) {
+	// Simple pure C++ implementation was faster than __builtin_clz version in this situation.
+	if (n < 10) return 1;
+	if (n < 100) return 2;
+	if (n < 1000) return 3;
+	if (n < 10000) return 4;
+	if (n < 100000) return 5;
+	if (n < 1000000) return 6;
+	if (n < 10000000) return 7;
+	if (n < 100000000) return 8;
+	if (n < 1000000000) return 9;
+	return 10;
+}
+
+static inline void DigitGen(const DiyFp W, const DiyFp Mp, uint64_t delta, char* buffer, int* len, int* K) {
+	static const uint32_t kPow10[] = { 1, 10, 100, 1000, 10000, 100000, 1000000, 10000000, 100000000, 1000000000 };
+	const DiyFp one = DiyFp_from_parts((uint64_t )(1) << -Mp.e, Mp.e);
+	const DiyFp wp_w = DiyFp_subtract(Mp, W);
+	uint32_t p1 = (uint32_t )(Mp.f >> -one.e);
+	uint64_t p2 = Mp.f & (one.f - 1);
+	int kappa = (int )(CountDecimalDigit32(p1));
+	*len = 0;
+
+	while (kappa > 0) {
+		uint32_t d;
+		switch (kappa) {
+			case 10: d = p1 / 1000000000; p1 %= 1000000000; break;
+			case  9: d = p1 /  100000000; p1 %=  100000000; break;
+			case  8: d = p1 /   10000000; p1 %=   10000000; break;
+			case  7: d = p1 /    1000000; p1 %=    1000000; break;
+			case  6: d = p1 /     100000; p1 %=     100000; break;
+			case  5: d = p1 /      10000; p1 %=      10000; break;
+			case  4: d = p1 /       1000; p1 %=       1000; break;
+			case  3: d = p1 /        100; p1 %=        100; break;
+			case  2: d = p1 /         10; p1 %=         10; break;
+			case  1: d = p1;              p1 =           0; break;
+			default: 
+#if defined(_MSC_VER)
+				__assume(0);
+#elif __GNUC__ > 4 || (__GNUC__ == 4 && __GNUC_MINOR__ >= 5)
+				__builtin_unreachable();
+#else
+				d = 0;
+#endif
+		}
+		if (d || *len)
+			buffer[(*len)++] = '0' + (char )(d);
+		kappa--;
+		uint64_t tmp = ((uint64_t )(p1) << -one.e) + p2;
+		if (tmp <= delta) {
+			*K += kappa;
+			GrisuRound(buffer, *len, delta, tmp, (uint64_t )(kPow10[kappa]) << -one.e, wp_w.f);
+			return;
+		}
+	}
+
+	// kappa = 0
+	for (;;) {
```

Comment:
Looking at [this paper's `Figure 6` line 8](https://www.cs.tufts.edu/~nr/cs257/archive/florian-loitsch/printf.pdf), I modified this line as below
```suggestion
	while ((*len) < 19) 
```
This should guarantee that there will be no buffer overruns. I see no lose of precision in round-trip test.

Reply:
I think it's a good idea to limit *len in that loop.  That's the one place where it's not obvious how many characters can be added to the buffer.

But as far as I understand the algorithm that limit should never be exceeded. So I think we should add `if ((*len) >= 19) abort();`. 

For our buffer limit, I think we should use a multiple of 8 since our stack is 8-byte aligned. If we use 40, it is very obvious that the buffer cannot overflow.